### PR TITLE
feat(route): add 中国作家网

### LINKED ISSUE #13229: 希望将 中国作家网-报刊在线 加入到订阅源
### 类型

阅读

### 网站地址

http://www.chinawriter.com.cn/404015/416204/418925/

### 网站描述

中国作家网-报刊在线 类目下有丰富的严肃文学类报刊杂志，如《人民文学》《十月》《收获》《北京文学》等

### 需要生成什么内容？

为杂志内的文章生成可全文阅读的RSS

### 额外描述

中国作家网-世界文坛 栏目下也有不错的文学评论的内容，希望也能收录到RssHub中
http://www.chinawriter.com.cn/404085/index.html

### 这不是重复的 RSS 请求

- [X] 我已经搜索了[现有 issue](https://github.com/DIYgod/RSSHub/issues) 和 [pull requests](https://github.com/DIYgod/RSSHub/pulls)，以确保该 RSS 尚未被请求。

### PR DESCRIPTION
**Pull request by nczitzk**
## Involved Issue / 该 PR 相关 Issue

Close #13229

## Example for the Proposed Route(s) / 路由地址示例

```routes
/chinawriter
/chinawriter/404085
/chinawriter/404015/416204/418925
```

## New RSS Route Checklist / 新 RSS 路由检查表
  
- [x] New Route / 新的路由
  - [x] Follows [v2 Script Standard](https://docs.rsshub.app/joinus/advanced/script-standard) / 跟随 [v2 路由规范](https://docs.rsshub.app/zh/joinus/advanced/script-standard)
- [x] Documentation / 文档说明
  - [ ] EN / 英文文档
  - [x] CN / 中文文档
- [x] Full text / 全文获取
  - [x] Use cache / 使用缓存
- [ ] Anti-bot or rate limit / 反爬/频率限制 
  - [ ] If yes, do your code reflect this sign? / 如果有, 是否有对应的措施? 
- [x] [Date and time](https://docs.rsshub.app/joinus/advanced/pub-date) / [日期和时间](https://docs.rsshub.app/zh/joinus/advanced/pub-date)
  - [x] Parsed / 可以解析
  - [x] Correct time zone / 时区正确
- [ ] New package added / 添加了新的包
- [ ] `Puppeteer`

### CODE REVIEW

File: lib/v2/chinawriter/index.js [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+                const content = cheerio.load(decodeBuffer(detailResponse));
+
+                content('div.end_shared').remove();
+
+                const info = content('div.end_info').text().trim();
+
+                item.title = content('#newstit, h6.end_tit').text();
+                item.description = content('div.end_article').html();
+                item.author = info.match(/\|(.*)\d{4}/)[1].trim();
```
Comment (TonyRL) [inline lib/v2/chinawriter/index.js:54]: This regex may not always works.
e.g., `http://www.chinawriter.com.cn/n1/2023/0303/c404928-32635716.html` from `/chinawriter/403978`

File: lib/v2/chinawriter/index.js [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+                item.category = [
+                    ...content('div.location a.clink')
+                        .slice(1)
+                        .toArray()
+                        .map((c) => content(c).text()),
+                    info.match(/来源：(.*)\|/)[1].trim(),
+                ];
+                item.pubDate = timezone(parseDate(content('div.end_info em').text(), 'YYYY年MM月DD日HH:mm'), +8);
```
Comment (TonyRL) [inline lib/v2/chinawriter/index.js:62]: You may consider adding $('meta[name="publishdate"]').attr('content') as last resort for pubDate for pages like `/chinawriter/403978`

File: lib/v2/chinawriter/index.js [hunk 3]
```diff
@@ -0,0 +1,83 @@
+const got = require('@/utils/got');
+const cheerio = require('cheerio');
+const iconv = require('iconv-lite');
+const jschardet = require('jschardet');
+const timezone = require('@/utils/timezone');
+const { parseDate } = require('@/utils/parse-date');
+
+const decodeBuffer = (buffer) => iconv.decode(buffer, jschardet.detect(buffer).encoding);
```
Comment (TonyRL) [inline lib/v2/chinawriter/index.js:8]: I've tested all the entries in the docs. All of them can be decoded properly using the default `responseType`. `iconv-lite` isn't needed for this site.

Beside that, you can load the response buffer as `utf-8` and check the attribute of `$('meta[charset]')` or `$('meta[http-equiv="Content-Type"]')` if the response header does not provide charset.

File: website/docs/routes/new-media.md [hunk 4]
```diff
@@ -5332,6 +5332,117 @@ column 为 third 时可选的 category:
...
 
+## 中国作家网 {#zhong-guo-zuo-jia-wang}
+
+### 栏目 {#zhong-guo-zuo-jia-wang-lan-mu}
+
+<Route author="nczitzk" example="/chinawriter" path="/chinawriter/:id?" paramsDesc={['栏目 id，见下表，默认为首页']} radar="1" rssbud="1">
+
+| 机构          | 服务   | 会员   | 文学奖项 |
+| ------------- | ------ | ------ | -------- |
+| 403936/419343 | 403937 | 403978 | 403973   |
```
[the remark this answers is not in the document]
Comment (nczitzk) [reply website/docs/routes/new-media.md:5343]: Thank you very much for such a detailed review. ❤️

File: lib/v2/chinawriter/index.js [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+                                info
+                                    ? info
+                                          .match(/^(.*)\|/)[1]
+                                          .replace(/来源：/g, '')
+                                          .trim()
+                                    : undefined,
+                            ].filter(Boolean)
+                        ),
+                    ];
+                    item.pubDate = content('div.end_info em').text() ? timezone(parseDate(content('div.end_info em').text(), 'YYYY年MM月DD日HH:mm'), +8) : parseDate(content('meta[name="description"]').prop('content'));
```
Comment (TonyRL) [inline lib/v2/chinawriter/index.js:64]:
```suggestion
                    item.pubDate = content('div.end_info em').text() ? timezone(parseDate(content('div.end_info em').text(), 'YYYY年MM月DD日HH:mm'), +8) : parseDate(content('meta[name="publishdate"]').prop('content'));
```
https://github.com/DIYgod/RSSHub/pull/13271#discussion_r1323445876

So `http://www.chinawriter.com.cn/n1/2023/0303/c404928-32635716.html` from `/chinawriter/403978` can have proper pubDate.

File: lib/v2/chinawriter/index.js [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+                } catch (e) {
+                    // eslint-disable-next-line
```
Comment (TonyRL) [inline lib/v2/chinawriter/index.js:66]:
```suggestion
                    //
```

File: lib/v2/chinawriter/index.js [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+const cheerio = require('cheerio');
+const timezone = require('@/utils/timezone');
+const { parseDate } = require('@/utils/parse-date');
+
+module.exports = async (ctx) => {
+    const { id } = ctx.params;
+    const limit = ctx.query.limit ? parseInt(ctx.query.limit, 10) : 40;
+
+    const rootUrl = 'http://www.chinawriter.com.cn';
+    const currentUrl = new URL(id ?? '', rootUrl).href;
```
Comment (TonyRL) [inline lib/v2/chinawriter/index.js:11]:
```suggestion
    const currentUrl = `${new URL(id ?? '', rootUrl).href}/`;
```
Use the url with trailing hash directly.
https://github.com/DIYgod/RSSHub/actions/runs/6174984526/job/16760843470#step:12:102

**Pull request merged**